### PR TITLE
Update mypy to 1.3.0 and fix type issues

### LINKED ISSUE #70: [Feature] Upgrade to mypy 1.3.0
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward change to existing dbt-semantic-interfaces functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

The version of mypy should be upgraded from `1.1.1` to the latest version `1.3.0` to be consistent with what's used in `metricflow`.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by plypaul**
resolves #70 

### Description

As per title.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -29,17 +29,9 @@ repos:
       language: system
 
   - repo: https://github.com/pre-commit/mirrors-mypy # configured via mypy.ini
-    rev: v0.942
+    rev: v1.3.0
     hooks:
     - id: mypy
       args: [--ignore-missing-imports,--show-error-codes]
```
Comment (tlento) [inline .pre-commit-config.yaml:35]: We'll need to remove --ignore-missing-imports at some point.

File: dbt_semantic_interfaces/dataclass_serialization.py [hunk 2]
```diff
@@ -221,11 +225,14 @@ def _convert_dataclass_instance_to_pydantic_model(
         return obj
 
     def pydantic_serialize(self, obj: SerializableDataclassT) -> str:  # noqa: D
-        assert dataclasses.is_dataclass(obj)
+        obj_class = obj.__class__
+        assert dataclasses.is_dataclass(obj), f"Got object of type: {obj.__class__.__name__}"
+        assert isinstance(obj, SerializableDataclass), f"Got object of type: {obj.__class__.__name__}"
+        assert issubclass(obj_class, SerializableDataclass), f"Got object type: {obj_class.__name__}"
```
Comment (tlento) [inline dbt_semantic_interfaces/dataclass_serialization.py:231]: Why not use the local obj_class in the messages?

File: dbt_semantic_interfaces/dataclass_serialization.py [hunk 3]
```diff
@@ -221,11 +225,14 @@ def _convert_dataclass_instance_to_pydantic_model(
...
 
         return self._convert_dataclass_instance_to_pydantic_model(
             # .__class__ seems to be the approach for new classes and there are differences with type(obj)
```
Comment (tlento) [inline dbt_semantic_interfaces/dataclass_serialization.py:234]: Move this comment to where the local is defined?

**Review by tlento — APPROVED** [review]
Sweet, thanks!

**Review by QMalcolm — COMMENTED** [review]
Thank you for doing this work. Being up to date with mypy and actually type checking stuff we were previously skipping is an improvement we absolutely needed to make. I do have some questions on the changes to `from_issues_sequence`

File: dbt_semantic_interfaces/validations/validator_helpers.py [hunk 4]
```diff
@@ -168,6 +167,38 @@ def as_cli_formatted_str(self, verbose: bool = False) -> str:
...
         )
 
+    @property
+    @abstractmethod
+    def as_issue_set(self) -> ValidationIssueSet:  # noqa: D
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class ValidationIssueSet:
```
[the remark this answers is not in the document]
Comment (plypaul) [reply dbt_semantic_interfaces/validations/validator_helpers.py:177]: Yeah, but I saw it had some additional methods and that seemed like a bigger change that was more apt for a follow up.

File: dbt_semantic_interfaces/validations/validator_helpers.py [hunk 5]
```diff
@@ -214,23 +257,15 @@ def has_blocking_issues(self) -> bool:
...
-                errors.append(issue)
-            else:
-                assert_values_exhausted(issue.level)
-        return cls(warnings=tuple(warnings), future_errors=tuple(future_errors), errors=tuple(errors))
+        combined_issue_set = ValidationIssueSet.combine(tuple(issue.as_issue_set for issue in issues))
+        return ModelValidationResults(
+            warnings=tuple(combined_issue_set.warning_issues),
+            future_errors=tuple(combined_issue_set.future_error_issues),
+            errors=tuple(combined_issue_set.error_issues),
+        )
```
[the remark this answers is not in the document]
Comment (plypaul) [reply dbt_semantic_interfaces/validations/validator_helpers.py:268]: I'm avoiding uses of `isinstance` in the codebase unless absolutely needed, like introspection.

File: dbt_semantic_interfaces/dataclass_serialization.py (hunk 2 above)
Comment (plypaul) [reply dbt_semantic_interfaces/dataclass_serialization.py]: Updated.

File: dbt_semantic_interfaces/dataclass_serialization.py (hunk 3 above)
Comment (plypaul) [reply dbt_semantic_interfaces/dataclass_serialization.py]: Updated.

**Pull request merged**